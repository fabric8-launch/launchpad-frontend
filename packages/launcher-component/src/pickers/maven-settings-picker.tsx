import { Form, FormGroup, TextInput } from '@patternfly/react-core';
import React from 'react';
import { InputProps, Picker } from '../core/types';

export interface MavenSettingsPickerValue {
  groupId: string;
  artifactId: string;
  version: string;
}

interface MavenSettingsPickerProps extends InputProps<MavenSettingsPickerValue> {
}

const VALUE_REGEXP = /^[a-z][a-z0-9-\.]{3,63}$/;

export const MavenSettingsPicker: Picker<MavenSettingsPickerProps, MavenSettingsPickerValue> = {
  checkCompletion: value => !!value.groupId && VALUE_REGEXP.test(value.groupId)
    && !!value.artifactId && VALUE_REGEXP.test(value.artifactId) && !!value.version,
  Element: props => {
    const isValid = (value: string) => VALUE_REGEXP.test(value || '');
    return (
      <Form>
        <FormGroup
          label="GroupId"
          isRequired
          fieldId="groupId"
          helperTextInvalid="Please provide a valid groupId"
          isValid={isValid(props.value.groupId)}
        >
          <TextInput
            isRequired
            type="text"
            id="groupId"
            name="groupId"
            aria-describedby="groupId"
            value={props.value.groupId}
            onChange={value => props.onChange({ ...props.value, groupId: value })}
            pattern={VALUE_REGEXP.source}
            isValid={isValid(props.value.groupId)}
          />
        </FormGroup>
        <FormGroup
          label="ArtifactId"
          isRequired
          fieldId="artifactId"
          helperTextInvalid="Please provide a valid artifactId"
          isValid={isValid(props.value.artifactId)}
        >
          <TextInput
            isRequired
            type="text"
            id="artifactId"
            name="artifactId"
            aria-describedby="artifactId"
            value={props.value.artifactId}
            onChange={value => props.onChange({ ...props.value, artifactId: value })}
            pattern={VALUE_REGEXP.source}
            isValid={isValid(props.value.artifactId)}
          />
        </FormGroup>
        <FormGroup
          label="Version"
          isRequired
          fieldId="version"
          helperTextInvalid="Please provide a version number"
          isValid={!!props.value.version}
        >
          <TextInput
            isRequired
            type="text"
            id="version"
            name="version"
            aria-describedby="version"
            value={props.value.version}
            onChange={value => props.onChange({ ...props.value, version: value })}
            isValid={!!props.value.version}
          />
        </FormGroup>
      </Form>
    );
  }
};
