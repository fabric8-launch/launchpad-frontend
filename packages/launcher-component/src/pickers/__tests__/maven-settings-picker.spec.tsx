import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { FormPanel } from '../../core/form-panel/form-panel';

import { MavenSettingsPicker } from '../maven-settings-picker';

afterEach(cleanup);

describe('<MavenSettingsPicker />', () => {
  it('renders the MavenSettingsPicker correctly', () => {
    const comp = render(<MavenSettingsPicker.Element value={{groupId: '', version: '', artifactId: ''}} onChange={() => {}}/>);
    expect(comp.asFragment()).toMatchSnapshot();
  });

  it('show error for invalid data', () => {
    const handleSave = jest.fn();
    const comp = render(
      <FormPanel
        initialValue={{}}
        validator={MavenSettingsPicker.checkCompletion}
        onSave={handleSave}
        onCancel={() => {}}
      >
        {(inputProps) => (<MavenSettingsPicker.Element {...inputProps} showMoreOptions />)}
      </FormPanel>
    );

    const groupIdField = comp.getByLabelText('Maven groupId name');
    fireEvent.change(groupIdField, { target: { value: 'invalid name' } });
    const artifactIdField = comp.getByLabelText('Maven artifactId name');
    fireEvent.change(artifactIdField, { target: { value: 'invalid name' } });
    const toggleButton = comp.getByLabelText('Expand panel');
    fireEvent.click(toggleButton);
    const versionField = comp.getByLabelText('Maven version number');
    fireEvent.change(versionField, { target: { value: '1' } });
    expect(comp.asFragment()).toMatchSnapshot();
  });
});
