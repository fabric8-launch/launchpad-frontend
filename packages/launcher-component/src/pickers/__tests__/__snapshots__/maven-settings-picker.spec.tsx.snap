// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MavenSettingsPicker /> renders the MavenSettingsPicker correctly 1`] = `
<DocumentFragment>
  <div
    class="maven-settings-picker vertical"
  >
    <div
      class="base-settings pf-c-form"
    >
      <div
        class="pf-c-form__group"
      >
        <label
          class="pf-c-form__label"
          for="groupId"
        >
          <span
            class="pf-c-form__label-text"
          >
            Group
          </span>
        </label>
        <input
          aria-invalid="false"
          aria-label="Maven groupId name"
          class="pf-c-form-control"
          id="groupId"
          name="groupId"
          pattern="^[a-z][a-z0-9-.]{3,63}$"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="pf-c-form__group"
      >
        <label
          class="pf-c-form__label"
          for="artifactId"
        >
          <span
            class="pf-c-form__label-text"
          >
            Artifact
          </span>
        </label>
        <input
          aria-invalid="false"
          aria-label="Maven artifactId name"
          class="pf-c-form-control"
          id="artifactId"
          name="artifactId"
          pattern="^[a-z][a-z0-9-.]{3,63}$"
          required=""
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="toggle-panel vertical "
    >
      <div
        class="extended-settings pf-c-form"
      >
        <div
          class="pf-c-form__group"
        >
          <label
            class="pf-c-form__label"
            for="version"
          >
            <span
              class="pf-c-form__label-text"
            >
              Version
            </span>
          </label>
          <input
            aria-invalid="false"
            aria-label="Maven version number"
            class="pf-c-form-control"
            id="version"
            name="version"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="pf-c-form__group"
        >
          <label
            class="pf-c-form__label"
            for="description"
          >
            <span
              class="pf-c-form__label-text"
            >
              Description
            </span>
          </label>
          <input
            aria-invalid="false"
            aria-label="Maven description"
            class="pf-c-form-control"
            id="description"
            name="description"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="pf-c-form__group"
        >
          <label
            class="pf-c-form__label"
            for="packageName"
          >
            <span
              class="pf-c-form__label-text"
            >
              Package Name
            </span>
          </label>
          <input
            aria-invalid="false"
            aria-label="Maven package name"
            class="pf-c-form-control"
            id="packageName"
            name="packageName"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="toggle-button"
    >
      <a
        aria-disabled="false"
        aria-label="Toggle panel"
        class="pf-c-button pf-m-link"
      >
        <span>
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 448 512"
            width="1em"
          >
            <path
              d="M448 80v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V80c0-26.5 21.5-48 48-48h352c26.5 0 48 21.5 48 48zM92.5 220.5l123 123c4.7 4.7 12.3 4.7 17 0l123-123c7.6-7.6 2.2-20.5-8.5-20.5H101c-10.7 0-16.1 12.9-8.5 20.5z"
              transform=""
            />
          </svg>
           Open
        </span>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<MavenSettingsPicker /> show error for invalid data 1`] = `
<DocumentFragment>
  <div
    class="form-panel"
    style="padding: 20px;"
  >
    <div
      class="maven-settings-picker vertical"
    >
      <div
        class="base-settings pf-c-form"
      >
        <div
          class="pf-c-form__group"
        >
          <label
            class="pf-c-form__label"
            for="groupId"
          >
            <span
              class="pf-c-form__label-text"
            >
              Group
            </span>
          </label>
          <input
            aria-invalid="true"
            aria-label="Maven groupId name"
            class="pf-c-form-control"
            id="groupId"
            name="groupId"
            pattern="^[a-z][a-z0-9-.]{3,63}$"
            required=""
            type="text"
            value="invalid name"
          />
          <div
            aria-live="polite"
            class="pf-c-form__helper-text pf-m-error"
            id="groupId-helper"
          >
            Please provide a valid groupId
          </div>
        </div>
        <div
          class="pf-c-form__group"
        >
          <label
            class="pf-c-form__label"
            for="artifactId"
          >
            <span
              class="pf-c-form__label-text"
            >
              Artifact
            </span>
          </label>
          <input
            aria-invalid="true"
            aria-label="Maven artifactId name"
            class="pf-c-form-control"
            id="artifactId"
            name="artifactId"
            pattern="^[a-z][a-z0-9-.]{3,63}$"
            required=""
            type="text"
            value="invalid name"
          />
          <div
            aria-live="polite"
            class="pf-c-form__helper-text pf-m-error"
            id="artifactId-helper"
          >
            Please provide a valid artifactId
          </div>
        </div>
      </div>
      <div
        class="toggle-panel vertical open"
      >
        <div
          class="extended-settings pf-c-form"
        >
          <div
            class="pf-c-form__group"
          >
            <label
              class="pf-c-form__label"
              for="version"
            >
              <span
                class="pf-c-form__label-text"
              >
                Version
              </span>
            </label>
            <input
              aria-invalid="false"
              aria-label="Maven version number"
              class="pf-c-form-control"
              id="version"
              name="version"
              required=""
              type="text"
              value="1"
            />
          </div>
          <div
            class="pf-c-form__group"
          >
            <label
              class="pf-c-form__label"
              for="description"
            >
              <span
                class="pf-c-form__label-text"
              >
                Description
              </span>
            </label>
            <input
              aria-invalid="false"
              aria-label="Maven description"
              class="pf-c-form-control"
              id="description"
              name="description"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="pf-c-form__group"
          >
            <label
              class="pf-c-form__label"
              for="packageName"
            >
              <span
                class="pf-c-form__label-text"
              >
                Package Name
              </span>
            </label>
            <input
              aria-invalid="false"
              aria-label="Maven package name"
              class="pf-c-form-control"
              id="packageName"
              name="packageName"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="toggle-button"
      >
        <a
          aria-disabled="false"
          aria-label="Toggle panel"
          class="pf-c-button pf-m-link"
        >
          <span>
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M0 432V80c0-26.51 21.49-48 48-48h352c26.51 0 48 21.49 48 48v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48zm355.515-140.485l-123.03-123.03c-4.686-4.686-12.284-4.686-16.971 0L92.485 291.515c-7.56 7.56-2.206 20.485 8.485 20.485h246.059c10.691 0 16.045-12.926 8.486-20.485z"
                transform=""
              />
            </svg>
             Close
          </span>
        </a>
      </div>
    </div>
    <hr
      class="separator"
    />
    <div
      class="pf-l-toolbar"
    >
      <div
        class="pf-l-toolbar__group"
      >
        <button
          aria-label="Save form"
          class="pf-c-button pf-m-primary pf-m-disabled"
          disabled=""
          type="button"
        >
          Save
        </button>
      </div>
      <div
        class="pf-l-toolbar__group"
      >
        <button
          aria-label="Cancel form"
          class="pf-c-button pf-m-secondary"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
